src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -29411,48 +29411,206 @@ public void M(string? s)
     {
         if (s is null) return;
         var t = M2(ref s);
-        s.ToString(); // 1
-        t.ToString(); // 2
+        s.ToString();
+        t.ToString();
     }
     T M2<T>(ref T t) => throw null!;
 }
 ");
 
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(37909, "https://github.com/dotnet/roslyn/issues/37909")]
+        public void OutParameter_TypeInferenceUsesLValueType()
+        {
+            var c = CreateNullableCompilation(@"
+class C
+{
+    void M()
+    {
+        string? s1;
+        var t1 = M2(out s1);
+        s1.ToString(); // 1
+        t1.ToString(); // 2
+
+        string? s2 = string.Empty;
+        var t2 = M2(out s2);
+        s2.ToString(); // 3
+        t2.ToString(); // 4
+
+        string? s3 = null;
+        var t3 = M2(out s3);
+        s3.ToString(); // 5
+        t3.ToString(); // 6
+    }
+
+    T M2<T>(out T t) => throw null!;
+}
+");
             c.VerifyDiagnostics(
                 // (8,9): warning CS8602: Dereference of a possibly null reference.
-                //         s.ToString(); // 1
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(8, 9),
+                //         s1.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s1").WithLocation(8, 9),
                 // (9,9): warning CS8602: Dereference of a possibly null reference.
-                //         t.ToString(); // 2
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t").WithLocation(9, 9)
+                //         t1.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t1").WithLocation(9, 9),
+                // (13,9): warning CS8602: Dereference of a possibly null reference.
+                //         s2.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s2").WithLocation(13, 9),
+                // (14,9): warning CS8602: Dereference of a possibly null reference.
+                //         t2.ToString(); // 4
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t2").WithLocation(14, 9),
+                // (18,9): warning CS8602: Dereference of a possibly null reference.
+                //         s3.ToString(); // 5
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s3").WithLocation(18, 9),
+                // (19,9): warning CS8602: Dereference of a possibly null reference.
+                //         t3.ToString(); // 6
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t3").WithLocation(19, 9)
                 );
         }
 
+        [Fact]
+        [WorkItem(47663, "https://github.com/dotnet/roslyn/issues/47663")]
+        public void RefParameter_Issue_47663()
+        {
+            var source = @"
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+class C
+{
+    static void X1<T>([NotNullIfNotNull(""value"")] ref T location, T value) where T : class? { }
```

Comment:
How did your change affect this scenario? I can get how before we inferred a `double[]` type argument and now we infer a `double[]?` type argument, but how did that result in the postcondition for the NotNullIfNotNullAttribute starting to be respected?

Reply:
We used to produce a warning on `ref f1` for assigning maybe-null value inbound to the first parameter (which was inferred as `double[]!`.
For `ref f2` we now produce a diagnostic on the outbound assignment from parameter `double[]?` to local typed `double[]!`.

---
In reply to: [493893383](https://github.com/dotnet/roslyn/pull/47952#discussion_r493893383) [](ancestors = 493893383)

Reply:
Makes complete sense now, thanks!